FLT/HIMExperiments/module_topology.lean

```diff
@@ -137,13 +136,81 @@ example (ι : Type*) [Finite ι] :
     -- it's linear, because Lean has the projections as linear maps.
     exact ⟨LinearMap.proj i, rfl⟩
 
+--maybe I should conclude this from the example above but don't know how
+--otherwise this should be very easy to do by hand
+lemma Module.topology_self : (hA : TopologicalSpace A) = Module.topology A := by
+  apply le_antisymm
+  · unfold topology
+    apply le_iInf
+    intro i
+    rw [← continuous_iff_le_induced]
+    have hi : i = LinearMap.lsmul A A (i 1) := by
+      ext
+      simp only [LinearMap.lsmul_apply, smul_eq_mul, mul_one]
+    rw [hi]
+    apply continuous_const.mul continuous_id
+    -- a linear map should always be continuous in this setting right?
+  · apply sInf_le
+    rw [Set.mem_range]
+    use LinearMap.id
+    exact induced_id
+
 -- We need that the module topology on a product is the product topology
 lemma Module.prod_canonical :
     @instTopologicalSpaceProd M N (Module.topology A) (Module.topology A) =
     (Module.topology A : TopologicalSpace (M × N)) := by
   -- the goal is to prove that an iInf equals an inf (of two topologies).
-  unfold instTopologicalSpaceProd -- you can probably delete this later
-  sorry
+  apply le_antisymm
+  · apply le_iInf
+    intro f
+    rw [← continuous_iff_le_induced]
+    have : ⇑f = fun ⟨m, n⟩ ↦ f (⟨m, 0⟩) + f (⟨0, n⟩) := by
```

Comment:
So far so good. In my `example` I now applied a lemma from the library, and you have decided to prove the analogous lemma *in the middle of another proof*. This is not normally a good strategy -- those lemmas should be pulled out. You should formalise the statement of what you need, then check with loogle and `exact?` to see if it's there, and if it's not there you should think of a good name (e.g. `LinearMap.continuous_on_prod`) and then write it separately outside this proof. Lots of short proofs is always better than one long one.